### PR TITLE
Reuse images across jobs

### LINKED ISSUE #425: Remove supervisord in ingestion server image
## Problem

Currently the ingestion server installs `supervisord` and uses it to run the integration server. Ostensibly, it looks like this was done in order to restart the ingestion server if it fell over.

## Description

Service restarts should be the purview of the Docker supervisor (e.g. docker-compose in this case) and should not be managed within the container. Additionally, the Docker command is always run with PID 1. Replacing this with `supervisord` means that any signals which get sent to the Docker container (and thus `supervisord`) may not be properly propagated to the webserver process.

As part of this, we should also set the relevant docker-compose files to restart the container automatically.

## Additional context

This appears to be the only context I can find for adding `supervisord`:
https://github.com/creativecommons/cccatalog-api/commit/4a13a40a7e4c94fea0a097799b6ded9db12315c3

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by dhruvkb**
## Description

This PR
- sets up a mechanism to reuse the images across jobs using GitHub Action artifacts
- uses multi-stage builds to drastically cut down the sizes of our Python images by a total of **82.07%** (see below)
- updates all images to Python 3.10
- removes `supervisord` from `ingestion_server` (fixes #425)
- unifies the `ingestion_server` and `indexer_worker` images 
- simplifies image build matrix configuration (conflict with #424)

To goal of the PR is to eventually unify with #424 in creating a comprehensive CI + CD pipeline that uses caching and artifacts to speedily test code, deploy images to GH CR and deploy docs to GH Pages. The presence of these artifacts can also make #213 easier.

## Image difference

| Name | Old | New |
|-|-|-|
| analytics | 1.24GB | 199MB | 
| ingestion_server | 1.12GB | 270MB |
| indexer_worker | 1.10GB | Gone! |
| openverse_api | 1.28GB | 401MB |
| Total | 4.74GB | 870MB |

## Testing Instructions

This PR makes changes to CI itself so all passing CI checks would be adequate. You can also see if all the services work just like before by following the steps in the `README.md`.

To see difference in image sizes,
- see current image sizes: `docker image ls`
- rebuild all images: `docker-compose build web ingestion_server analytics`
- see new image sizes: `docker image ls`
- compute, and optionally appreciate, the difference

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by dhruvkb** [comment]
Considering that all three `Dockerfile`s have *very* similar intermediate `builder` images, it might be possible to speed up the builds by unifying the `builder` image in all `Dockerfile`s. Trading just a bit of the size improvements to install `librdkafka-dev` in all three repos would mean that they all use the same builder!

Not sure how it would be implemented, just putting this idea out there.

**Review by sarayourfriend — APPROVED** [review]
This is amazing! CI is passing which is good but I also tested locally by doing `just recreate && just init && just stats` and it's working perfectly 👍 Nice work getting all of this together!

**Review by krysal — APPROVED** [review]
Awesomw x 💯 Seems that it even made the execution of jobs faster. Nice!

**Review by AetherUnbound — APPROVED** [review]
This is seriously amazing Dhruv!! I love it, way to cut down on image sizes :100: I was able to build and test these locally :rocket:

File: justfile [hunk 1]
```diff
@@ -67,6 +67,7 @@ init: up wait-for-es wait-for-ing wait-for-web
 @install:
     just _api-install
     just _ing-install
+    just _nl-install
```
[the remark this answers is not in the document]
Comment (dhruvkb) [reply justfile:71]: I'd prefer to make a separate smaller PR for this.

**Pull request merged**